tpu-client-next/src/connection_worker.rs

```diff
@@ -0,0 +1,231 @@
+//! functionality needed to handle one connection within the scope of task.
+use {
+    super::SendTransactionStats,
+    crate::{
+        quic_networking::send_data_over_stream, send_transaction_stats::record_error,
+        transaction_batch::TransactionBatch,
+    },
+    log::*,
+    quinn::{ConnectError, Connection, Endpoint},
+    solana_measure::measure::Measure,
+    solana_sdk::{
+        clock::{DEFAULT_MS_PER_SLOT, MAX_PROCESSING_AGE, NUM_CONSECUTIVE_LEADER_SLOTS},
+        timing::timestamp,
+    },
+    std::net::SocketAddr,
+    tokio::{
+        sync::mpsc,
+        time::{sleep, Duration},
+    },
+    tokio_util::sync::CancellationToken,
+};
+
+/// Maximum number of reconnection attempts in case the connection errors out.
+const MAX_RECONNECT_ATTEMPTS: usize = 4;
+
+/// Interval between retry attempts for creating a new connection. This value is
+/// a best-effort estimate, based on current network conditions.
+const RETRY_SLEEP_INTERVAL: Duration =
+    Duration::from_millis(NUM_CONSECUTIVE_LEADER_SLOTS * DEFAULT_MS_PER_SLOT);
+
+/// Maximum age (in milliseconds) of a blockhash, beyond which transaction
+/// batches are dropped.
+const MAX_PROCESSING_AGE_MS: u64 = MAX_PROCESSING_AGE as u64 * DEFAULT_MS_PER_SLOT;
+
+/// [`ConnectionState`] represents the current state of a quic connection. This
+/// enum tracks the lifecycle of connection from initial setup to closing phase.
+/// The transition function between states is defined in `ConnectionWorker`
+/// implementation.
+enum ConnectionState {
+    NotSetup,
+    Active(Connection),
+    Retry(usize),
+    Closing,
+}
+
+impl Drop for ConnectionState {
+    fn drop(&mut self) {
+        if let Self::Active(connection) = self {
+            debug!(
+                "Close connection with {:?}, stats: {:?}. All pending streams will be dropped.",
+                connection.remote_address(),
+                connection.stats()
+            );
+            // no guarantee that all the streams will finish
+            connection.close(0u32.into(), b"done");
+        }
+    }
+}
+
+/// [`ConnectionWorker`] holds connection to the validator with address `peer`. If
+/// connection has been closed, [`ConnectionWorker`] tries to reconnect
+/// [`MAX_RECONNECT_ATTEMPTS`] times. If connection is in `Active` state, it sends
+/// transactions received from `transactions_receiver`. Additionally, it
+/// accumulates statistics about connections and streams failures.
+pub(crate) struct ConnectionWorker {
+    endpoint: Endpoint,
+    peer: SocketAddr,
+    cancel: CancellationToken,
+    transactions_receiver: mpsc::Receiver<TransactionBatch>,
+    connection: ConnectionState,
+    send_txs_stats: SendTransactionStats,
+}
+
+impl ConnectionWorker {
+    pub fn new(
+        endpoint: Endpoint,
+        peer: SocketAddr,
+        transactions_receiver: mpsc::Receiver<TransactionBatch>,
+    ) -> (Self, CancellationToken) {
+        let cancel = CancellationToken::new();
```

Comment:
We could create a child token from the one received in scheduler and this way don't need to cancel explicitly in the shutdown. Not sure if worth the effort